src/nf/nf_batchnorm_layer_submodule.f90

```diff
@@ -0,0 +1,105 @@
+submodule(nf_batchnorm_layer) nf_batchnorm_layer_submodule
+
+  implicit none
+
+contains
+
+  pure module function batchnorm_layer_cons(num_features) result(res)
+    implicit none
+    integer, intent(in) :: num_features
+    type(batchnorm_layer) :: res
+
+    res % num_features = num_features
+    allocate(res % gamma(num_features), source=1.0)
+    allocate(res % beta(num_features))
+    allocate(res % running_mean(num_features), source=0.0)
+    allocate(res % running_var(num_features), source=1.0)
+    allocate(res % input(num_features, num_features))
+    allocate(res % output(num_features, num_features))
```

Comment:
I don't think the shape for these should be `(num_features, num_features)`, but rather `(batch_size, num_features)`. The `batch_size` also won't be known until the first forward pass, so we should defer the allocation until then. In the forward pass, we could have a simple `allocated` check to see if they have not been allocated then, and allocate them to the shape of the `input`.

Reply:
Sorry, I meant `(num_features, batch_size)`.